atomsci/ddm/pipeline/model_pipeline.py

```diff
@@ -42,40 +56,53 @@ def calc_AD_kmean_dist(train_dset, pred_dset, k, train_dset_pair_distance=None,
         # calculate the pairwise distance of training set
         train_dset_pair_distance = pairwise_distances(X=train_dset, metric=dist_metric)
     train_kmean_dis = []
+    # here we use k+1 since the nearest neighbor is itself.
+    kn_idxes = get_k_nearest_index(train_dset_pair_distance, k+1)
     for i in range(len(train_dset_pair_distance)):
-        kn_idx = np.argpartition(train_dset_pair_distance[i], k+1)
-        dis = np.mean(train_dset_pair_distance[i][kn_idx[:k+1]])
+        # instead of taking the mean, we divide by k. Leaving out the distance to itself
+        dis = np.sum(train_dset_pair_distance[i][kn_idxes[i]])/k
         train_kmean_dis.append(dis)
     train_dset_distribution = sp.stats.norm.fit(train_kmean_dis)
     # pairwise distance between train and pred set
     pred_size = len(pred_dset)
     train_pred_dis = pairwise_distances(X=pred_dset, Y=train_dset, metric=dist_metric)
+    train_pred_kn_idxes = get_k_nearest_index(train_pred_dis, k)
     pred_kmean_dis_score = np.zeros(pred_size)
     for i in range(pred_size):
-        pred_km_dis = np.mean(np.sort(train_pred_dis[i])[:k])
+        dists = train_pred_dis[i][train_pred_kn_idxes[i]]
+        pred_km_dis = np.mean(dists)
         train_dset_std = train_dset_distribution[1] if train_dset_distribution[1] != 0 else 1e-6
```

Comment:
should we keep this as 1e-6 or throw an error that there is no variability in the training dataset

Reply:
We have collisions of ECFP features that could cause this. I think throwing an error is kind of extreme, since I would want the rest of the AD indexes even if a few fail.

Maybe we should return NaN instead of 1e-6? I don't know why we didn't just return 0? Is this used downstream in AMPL anywhere or is it just for analysis?